File: 139-word-break/word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<int> dp(s.size(),0);

        for(int i=0;i<s.size();i++){
            for(auto w:wordDict){
                if(i<w.length()-1) continue;
                if(i==w.length()-1 || dp[i-w.length()]){
                    if(s.substr(i-w.length()+1,w.length())==w){
                        dp[i]=1;
                    }
                }
                
            }
        }
        return dp[s.size()-1];
        
    }
};

File: 139-word-break/word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[n];

        for (int i = 0; i < n; i++) {
            for (String word : wordDict) {
                if (i < word.length() - 1) continue;
                if (i == word.length() - 1 || dp[i - word.length()]) {
                    String substr = s.substring(i - word.length() + 1, i + 1);
                    if (substr.equals(word)) {
                        dp[i] = true;
                    }
                }
            }
        }
        return dp[n - 1];
    }
}